### PR TITLE
message stream impl for review

### PR DESCRIPTION
**Pull request by songtong**

### CODE REVIEW

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/build/ComponentsConfigurator.java [hunk 1]
```diff
@@ -65,16 +67,27 @@
...
 		      .req(BizLogger.class) //
 		);
-		all.add(C(CommandProcessor.class, CommandType.QUERY_OFFSET.toString(), QueryOffsetCommandProcessor.class)//
+		all.add(C(CommandProcessor.class, CommandType.QUERY_CONSUME_OFFSET.toString(), QueryConsumerOffsetCommandProcessor.class)//
 		      .req(BrokerLeaseContainer.class)//
 		      .req(BrokerConfig.class)//
 		      .req(MetaService.class)//
 		      .req(MessageQueueManager.class)//
 		);
+		all.add(C(CommandProcessor.class, CommandType.QUERY_MESSAGE_OFFSET.toString(), QueryMessageOffsetCommandProcessor.class)//
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/build/ComponentsConfigurator.java]: QUERY_OFFSET_BY_TIME

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/build/ComponentsConfigurator.java [hunk 2]
```diff
@@ -65,16 +67,27 @@
...
+		      PullSpecificMessageCommandProcessor.class)//
+		      .req(MetaService.class)//
+		      .req(MessageQueueManager.class)//
+		);
 		all.add(C(CommandProcessor.class, CommandType.MESSAGE_ACK.toString(), AckMessageCommandProcessor.class)//
 		      .req(MessageQueueManager.class)//
 		      .req(BizLogger.class) //
 		);
-		all.add(C(CommandProcessor.class, CommandType.QUERY_OFFSET.toString(), QueryOffsetCommandProcessor.class)//
+		all.add(C(CommandProcessor.class, CommandType.QUERY_CONSUME_OFFSET.toString(), QueryConsumerOffsetCommandProcessor.class)//
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/build/ComponentsConfigurator.java]: QUERY_LATEST_OFFSET

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/DefaultMessageQueue.java [hunk 3]
```diff
@@ -84,4 +88,18 @@ public Offset findLatestOffset(String groupId) {
 	private long getMaxOffset(long offset, AckHolder<MessageMeta> holder) {
 		return holder == null ? offset : Math.max(holder.getMaxAckedOffset(), offset);
 	}
+
+	@Override
+	public Offset findMessageOffset(long time) {
+		MessagePriority pmsg = (MessagePriority) m_storage.findMessageByTime(new Tpp(m_topic, m_partition, true), time);
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/DefaultMessageQueue.java]: select offset only, exclude blob etc.

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/DefaultMessageQueueManager.java [hunk 4]
```diff
@@ -232,11 +232,23 @@ public void initialize() throws InitializationException {
 	}
 
 	@Override
-	public Offset findLatestOffset(Tpg tpg) {
+	public Offset findLatestConsumeOffset(Tpg tpg) {
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/DefaultMessageQueueManager.java]: method name

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/DefaultMessageQueueManager.java [hunk 5]
```diff
@@ -232,11 +232,23 @@ public void initialize() throws InitializationException {
...
+	@Override
+	public Offset findMessageOffset(String topic, int partition, long time) {
+		if (!m_stopped.get()) {
+			return getMessageQueue(topic, partition).findMessageOffset(time);
+		}
+		return null;
+	}
+
+	@Override
+	public MessageQueue getQueue(String topic, int partitionId) {
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/DefaultMessageQueueManager.java]: ?

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/MessageQueueManager.java [hunk 6]
```diff
@@ -23,6 +23,8 @@
 
 	public MessageQueueCursor getCursor(Tpg tpg, Lease lease);
 
+	public MessageQueue getQueue(String topic, int partitionId);
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/MessageQueueManager.java]: ?

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java [hunk 7]
```diff
@@ -185,17 +185,48 @@ public synchronized Object findLastOffset(Tpp tpp, int groupId) throws Exception
...
+		List<MessagePriority> msgs = new ArrayList<MessagePriority>();
+		for (Long[] subOffsets : splitOffsets(offsets)) {
+			try {
+				msgs.addAll(m_msgDao.findWithOffsets(subOffsets, MessagePriorityEntity.READSET_FULL));
+			} catch (DalException e) {
+				log.error("Failed to fetch message({}).", tpp, e);
+				continue;
+			}
+		}
+		return prepareFetchResult(tpp, msgs);
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java]: creatFetchResult

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java [hunk 8]
```diff
@@ -501,4 +529,95 @@ private boolean resendAfter(ResendGroupId l, ResendGroupId r) {
...
 
+	@Override
+	public Object findMessageByTime(Tpp tpp, long time) {
+		return needFirstMessage(time) ? getFirstMessage(tpp) : //
+		      needLastMessage(time) ? getLastMessage(tpp) : //
+		            findMessageInRange(tpp, getFirstMessage(tpp), getLastMessage(tpp), time);
+	}
+
+	private boolean needFirstMessage(long time) {
+		return Long.MIN_VALUE == time || -2 == time;
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java]: remove -2

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java [hunk 9]
```diff
@@ -501,4 +529,95 @@ private boolean resendAfter(ResendGroupId l, ResendGroupId r) {
...
+	}
+
+	private boolean needLastMessage(long time) {
+		return Long.MAX_VALUE == time || -1 == time;
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java]: remove -1

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java [hunk 10]
```diff
@@ -501,4 +529,95 @@ private boolean resendAfter(ResendGroupId l, ResendGroupId r) {
 		return false;
 	}
 
+	@Override
+	public Object findMessageByTime(Tpp tpp, long time) {
+		return needFirstMessage(time) ? getFirstMessage(tpp) : //
+		      needLastMessage(time) ? getLastMessage(tpp) : //
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java]: use oldest and latest

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java [hunk 11]
```diff
@@ -501,4 +529,95 @@ private boolean resendAfter(ResendGroupId l, ResendGroupId r) {
...
+
+	private boolean needFirstMessage(long time) {
+		return Long.MIN_VALUE == time || -2 == time;
+	}
+
+	private boolean needLastMessage(long time) {
+		return Long.MAX_VALUE == time || -1 == time;
+	}
+
+	private MessagePriority findMessageInRange(Tpp tpp, MessagePriority firstMsg, MessagePriority lastMsg, long time) {
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java]: left, right just use <id, time> pair is enough

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java [hunk 12]
```diff
@@ -501,4 +529,95 @@ private boolean resendAfter(ResendGroupId l, ResendGroupId r) {
...
+		if (firstMsg == null //
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java]: use iteration

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java [hunk 13]
```diff
@@ -501,4 +529,95 @@ private boolean resendAfter(ResendGroupId l, ResendGroupId r) {
...
+				return msg;
+			default:
+				throw new RuntimeException("Impossiable case when find message by time!");
+			}
+		}
+
+		return null;
+	}
+
+	private MessagePriority findMessageById(Tpp tpp, long id) {
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java]: return <id, time> pair

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/transport/command/processor/PullSpecificMessageCommandProcessor.java [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+					if (!batches.isEmpty()) {
+						PullMessageResultCommandV2 cmd = new PullMessageResultCommandV2();
+						cmd.addBatches(batches);
+						cmd.getHeader().setCorrelationId(correlationId);
+						ctx.getChannel().writeAndFlush(cmd);
+						return;
+					}
+				}
+			} else {
+				logDebug("No broker lease to handle client pull message reqeust.", correlationId, tpg, null);
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/transport/command/processor/PullSpecificMessageCommandProcessor.java]: wrong debug message

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/transport/command/processor/PullSpecificMessageCommandProcessor.java [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+			}
+		} catch (Exception e) {
+			logDebug("Exception occurred while handling client pull message reqeust.", correlationId, tpg, e);
+		}
+		Command cmd = new PullMessageResultCommandV2();
```
Comment (leoliang) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/transport/command/processor/PullSpecificMessageCommandProcessor.java]: use another command to support acceptance result

File: hermes-broker/src/main/resources/META-INF/dal/jdbc/hermes-dal.xml [hunk 16]
```diff
@@ -25,6 +26,15 @@
         		LIMIT ${batch-size}]]>
 				</statement>
 			</query>
+			<query name="find-with-offsets" type="SELECT" multiple="true">
+				<param name="offsets" />
+				<statement><![CDATA[
+				SELECT <FIELDS/>
```
Comment (leoliang) [inline hermes-broker/src/main/resources/META-INF/dal/jdbc/hermes-dal.xml]: exclude unused fields

File: hermes-consumer/src/main/java/com/ctrip/hermes/consumer/DefaultConsumer.java [hunk 17]
```diff
@@ -22,6 +31,12 @@
 	@Inject
 	private Engine m_engine;
 
+	@Inject
+	private MessageStreamManager m_streamManager;
+
+	@Inject
+	private MetaService m_metaService;
```
Comment (leoliang) [inline hermes-consumer/src/main/java/com/ctrip/hermes/consumer/DefaultConsumer.java]: remove unused injection

File: hermes-consumer/src/main/java/com/ctrip/hermes/consumer/api/Consumer.java [hunk 18]
```diff
@@ -1,17 +1,25 @@
...
 	public abstract ConsumerHolder start(String topic, String groupId, MessageListener<?> listener);
-	
-	public abstract ConsumerHolder start(String topicPattern, String groupId, MessageListener<?> listener, MessageListenerConfig config);
-	
-	public abstract <T> List<MessageStream<T>> createMessageStreams(String topic, String groupId);
 
+	public abstract ConsumerHolder start(String topicPattern, String groupId, MessageListener<?> listener,
+	      MessageListenerConfig config);
+
+	public abstract <T> List<HermesMessageStream<T>> openMessageStreams(String topic, String groupId,
```
Comment (leoliang) [inline hermes-consumer/src/main/java/com/ctrip/hermes/consumer/api/Consumer.java]: return wrapper<list<T> stream, consumerholder>

File: hermes-consumer/src/main/java/com/ctrip/hermes/consumer/api/OffsetStorage.java [hunk 19]
```diff
@@ -0,0 +1,7 @@
+package com.ctrip.hermes.consumer.api;
+
+public interface OffsetStorage {
+	public MessageStreamOffset queryLatestOffset(String topic, int partitionId);
+
+	public void persistOffset(String topic, int partitionId, MessageStreamOffset offset);
```
Comment (leoliang) [inline hermes-consumer/src/main/java/com/ctrip/hermes/consumer/api/OffsetStorage.java]: remove this method?

File: hermes-consumer/src/main/java/com/ctrip/hermes/consumer/engine/bootstrap/strategy/StreamConsumingStrategyConsumerTask.java [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+
+		while (!isClosed() && !Thread.currentThread().isInterrupted() && !m_lease.get().isExpired()) {
+			try {
+				String topic = m_context.getTopic().getName();
+				MessageStreamOffset offset = m_context.getOffsetStorage().queryLatestOffset(topic, m_partitionId);
+				if (offset == null) {
+					queryOffsetSchedulePolicy.fail(true);
+					continue;
+				}
+				m_offset.set(new Offset(offset.getPriorityOffset(), offset.getNonPriorityOffset(), null));
```
Comment (leoliang) [inline hermes-consumer/src/main/java/com/ctrip/hermes/consumer/engine/bootstrap/strategy/StreamConsumingStrategyConsumerTask.java]: return if success

File: hermes-consumer/src/main/java/com/ctrip/hermes/consumer/stream/DefaultMessageStream.java [hunk 21]
```diff
@@ -0,0 +1,238 @@
...
+		}
+
+		@Override
+		public void remove() {
+			throw new UnsupportedOperationException("Can not remove item from message stream.");
+		}
+	}
+
+	@Override
+	public List<ConsumerMessage<T>> fetchMessages(Offset offset, int size) throws Exception {
```
Comment (leoliang) [inline hermes-consumer/src/main/java/com/ctrip/hermes/consumer/stream/DefaultMessageStream.java]: add parameter timeout

File: hermes-consumer/src/main/java/com/ctrip/hermes/consumer/stream/DefaultMessageStream.java [hunk 22]
```diff
@@ -0,0 +1,238 @@
...
+		public boolean hasNext() {
+			return getQueue().size() > 0;
+		}
+
+		@Override
+		public T next() {
+			try {
+				return getQueue().take();
+			} catch (InterruptedException e) {
+				log.error("Can not take message from queue.", e);
```
Comment (leoliang) [inline hermes-consumer/src/main/java/com/ctrip/hermes/consumer/stream/DefaultMessageStream.java]: Thread.currentThread().interrupt();

File: hermes-consumer/src/main/java/com/ctrip/hermes/consumer/stream/DefaultMessageStream.java [hunk 23]
```diff
@@ -0,0 +1,238 @@
...
+			Type[] genericInterfaces = m_iterator.getClass().getGenericInterfaces();
+			Type[] actualTypeArguments = ((ParameterizedType) genericInterfaces[0]).getActualTypeArguments();
+			@SuppressWarnings("rawtypes")
+			Class type = (Class) actualTypeArguments[0];
+			return type;
+		}
+	}
+
+	@Override
+	public List<ConsumerMessage<T>> fetchMessages(List<Offset> offsets) {
```
Comment (leoliang) [inline hermes-consumer/src/main/java/com/ctrip/hermes/consumer/stream/DefaultMessageStream.java]: remove this method

File: hermes-consumer/src/main/java/com/ctrip/hermes/consumer/stream/DefaultMessageStreamManager.java [hunk 24]
```diff
@@ -0,0 +1,107 @@
...
+
+	private Map<Pair<String, String>, ConsumerHolder> m_holders = new HashMap<>();
+
+	private Map<Pair<String, String>, StreamMessageListener<?>> m_listeners = new HashMap<>();
+
+	private Map<Pair<String, String>, List<HermesMessageStream<?>>> m_streams = new HashMap<>();
+
+	@Override
+	@SuppressWarnings({ "unchecked" })
+	public synchronized <T> List<HermesMessageStream<T>> openMessageStreams(String topicName, String groupId,
```
Comment (leoliang) [inline hermes-consumer/src/main/java/com/ctrip/hermes/consumer/stream/DefaultMessageStreamManager.java]: no singletone

File: hermes-consumer/src/main/java/com/ctrip/hermes/consumer/stream/StreamMessageListener.java [hunk 25]
```diff
@@ -0,0 +1,87 @@
...
+import com.ctrip.hermes.core.message.PropertiesHolderAware;
+import com.ctrip.hermes.core.message.ConsumerMessage.MessageStatus;
+import com.ctrip.hermes.meta.entity.Partition;
+import com.dianping.cat.Cat;
+import com.dianping.cat.configuration.NetworkInterfaceManager;
+import com.dianping.cat.message.Event;
+import com.dianping.cat.message.Transaction;
+import com.dianping.cat.message.spi.MessageTree;
+
+public class StreamMessageListener<T> implements MessageListener<T> {
```
Comment (leoliang) [inline hermes-consumer/src/main/java/com/ctrip/hermes/consumer/stream/StreamMessageListener.java]: extends BaseMessageListener, then you don't need to implement the batch onMessage method

File: hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java (hunk 10 above)
Comment (yiming187) [inline hermes-broker/src/main/java/com/ctrip/hermes/broker/queue/storage/mysql/MySQLMessageQueueStorage.java]: earliest and latest is better.

**Pull request closed without merging**